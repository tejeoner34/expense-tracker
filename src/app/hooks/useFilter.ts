'use client';

import { useState } from 'react';
import { useQuery } from '@tanstack/react-query';

type UseFilterProps = {
  userId: string;
  initialYear: number;
};

export function useFilter({ userId, initialYear }: UseFilterProps) {
  const [selectedYear, setSelectedYear] = useState(initialYear);

  const {
    data: expenses = [],
    isLoading,
    isError,
    refetch,
  } = useQuery({
    queryKey: ['expenses', userId, selectedYear],
    queryFn: async () => {
      const res = await fetch(`/api/expenses?userId=${userId}&year=${selectedYear}`);
      if (!res.ok) throw new Error('Failed to fetch expenses');
      return res.json();
    },
    refetchOnWindowFocus: false,
  });

  return {
    selectedYear,
    setSelectedYear,
    expenses,
    isLoading,
    isError,
    refetch,
  };
}
