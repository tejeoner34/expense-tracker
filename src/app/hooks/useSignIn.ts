import { useMutation } from '@tanstack/react-query';
import { useRouter } from 'next/navigation';
import { signIn } from 'next-auth/react';
import { routes } from '../routes/routes';

export function useSignIn() {
  const router = useRouter();
  const signInMutation = useMutation({
    mutationFn: async ({ email, password }: { email: string; password: string }) => {
      return await signIn('credentials', {
        email,
        password,
        redirect: false,
      });
    },
    onSuccess: (data) => {
      if (!data?.ok) {
        throw new Error(data?.error || 'Invalid login');
      }
      router.push(routes.dashboard);
    },
    onError: (error) => {
      console.log('error', error);
    },
  });
  return {
    signIn: signInMutation.mutate,
    isLoading: signInMutation.isPending,
    error: signInMutation.error,
  };
}
