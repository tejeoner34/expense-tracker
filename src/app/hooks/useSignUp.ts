import { Currency } from '@/domain/models/expense.model';
import { useMutation } from '@tanstack/react-query';
import { useRouter } from 'next/navigation';

export function useSignUp() {
  const router = useRouter();

  const signUpMutation = useMutation({
    mutationFn: async ({
      email,
      password,
      defaultCurrency,
    }: {
      email: string;
      password: string;
      defaultCurrency: Currency;
    }) => await signUp(email, password, defaultCurrency),
    onSuccess: () => {
      router.push('/sign-in');
    },
  });

  return {
    signUp: signUpMutation.mutate,
    isLoading: signUpMutation.isPending,
    error: signUpMutation.error,
  };
}

async function signUp(email: string, password: string, defaultCurrency: Currency) {
  const response = await fetch('/api/user', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ email, password, defaultCurrency }),
  });

  if (!response.ok) {
    const errorData = await response.json();
    throw new Error(errorData.message || response.statusText);
  }

  return await response.json();
}
