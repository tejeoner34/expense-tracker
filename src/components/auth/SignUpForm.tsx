'use client';

import { useForm } from 'react-hook-form';
import { Form, FormControl, FormField, FormItem, FormLabel, FormMessage } from '../ui/form';
import * as z from 'zod';
import { zodResolver } from '@hookform/resolvers/zod';
import { Input } from '../ui/input';
import { Button } from '../ui/button';
import Link from 'next/link';
import { useSignUp } from '@/app/hooks/useSignUp';
import { userSchemaSignUp } from '@/domain/schemas/user.schema';
import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from '../ui/select';
import { currencies } from '@/data/currencies';
import { Currency } from '@/domain/models/expense.model';

const SignUpForm = () => {
  const { signUp, error, isLoading } = useSignUp();
  const form = useForm<z.infer<typeof userSchemaSignUp>>({
    resolver: zodResolver(userSchemaSignUp),
    defaultValues: {
      email: '',
      password: '',
      defaultCurrency: currencies[0] as Currency,
    },
  });

  const onSubmit = async (values: z.infer<typeof userSchemaSignUp>) => {
    signUp(values);
  };

  return (
    <Form {...form}>
      <form onSubmit={form.handleSubmit(onSubmit)} className="w-full">
        <div className="space-y-2">
          <FormField
            control={form.control}
            name="email"
            render={({ field }) => (
              <FormItem>
                <FormLabel>Email</FormLabel>
                <FormControl>
                  <Input placeholder="mail@example.com" {...field} />
                </FormControl>
                <FormMessage />
              </FormItem>
            )}
          />
          <FormField
            control={form.control}
            name="password"
            render={({ field }) => (
              <FormItem>
                <FormLabel>Password</FormLabel>
                <FormControl>
                  <Input type="password" placeholder="Enter your password" {...field} />
                </FormControl>
                <FormMessage />
              </FormItem>
            )}
          />
          <FormField
            control={form.control}
            name="defaultCurrency"
            render={({ field }) => (
              <FormItem>
                <FormLabel>Default Currency</FormLabel>
                <FormControl>
                  <Select value={field.value} onValueChange={field.onChange}>
                    <SelectTrigger className="w-full">
                      <SelectValue placeholder="Currency" />
                    </SelectTrigger>
                    <SelectContent>
                      {currencies.map((currency) => (
                        <SelectItem key={currency} value={currency}>
                          {currency}
                        </SelectItem>
                      ))}
                    </SelectContent>
                  </Select>
                </FormControl>
                <FormMessage />
              </FormItem>
            )}
          />
        </div>
        <Button className="w-full mt-6" type="submit" disabled={isLoading}>
          Sign up {isLoading && '...'}
        </Button>
        {error && <span className="text-red-500">{error.message}</span>}
      </form>
      <div className="mx-auto my-4 flex w-full items-center justify-evenly before:mr-4 before:block before:h-px before:flex-grow before:bg-stone-400 after:ml-4 after:block after:h-px after:flex-grow after:bg-stone-400">
        or
      </div>
      <p className="text-center text-sm text-gray-600 mt-2">
        If you already have an account, please&nbsp;
        <Link className="text-blue-500 hover:underline" href="/sign-in">
          Sign in
        </Link>
      </p>
    </Form>
  );
};

export default SignUpForm;
