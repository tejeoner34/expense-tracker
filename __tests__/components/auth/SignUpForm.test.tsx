import '@testing-library/jest-dom';
import { render, screen, fireEvent, act } from '@testing-library/react';
import SignUpForm from '@/components/auth/SignUpForm';
import { useSignUp } from '@/app/hooks/useSignUp';

// Mock useSignUp
jest.mock('@/app/hooks/useSignUp', () => ({
  useSignUp: jest.fn().mockReturnValue({
    signUp: jest.fn(),
    error: null,
    isLoading: false,
  }),
}));

describe('SignUpForm', () => {
  it('renders email and password fields', () => {
    render(<SignUpForm />);

    expect(screen.getByLabelText(/email/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/password/i)).toBeInTheDocument();
  });

  it('displays validation errors when fields are empty', async () => {
    render(<SignUpForm />);

    fireEvent.submit(screen.getByRole('button', { name: /sign up/i }));

    expect(await screen.findByText(/email is required/i)).toBeInTheDocument();
    expect(await screen.findByText(/password is required/i)).toBeInTheDocument();
  });

  it('calls signUp function when form is submitted with valid data', async () => {
    const mockSignUp = jest.fn();
    (useSignUp as jest.Mock).mockReturnValue({
      signUp: mockSignUp,
      error: null,
      isLoading: false,
    });
    render(<SignUpForm />);

    await act(async () => {
      fireEvent.change(screen.getByLabelText(/email/i), {
        target: { value: 'test@example.com' },
      });
    });

    await act(async () => {
      fireEvent.change(screen.getByLabelText(/password/i), {
        target: { value: 'password123' },
      });
    });

    await act(async () => {
      fireEvent.submit(screen.getByRole('button', { name: /sign up/i }));
    });

    expect(mockSignUp).toHaveBeenCalledWith({
      email: 'test@example.com',
      password: 'password123',
    });
  });

  it('should match the snapshot', () => {
    const { asFragment } = render(<SignUpForm />);
    expect(asFragment()).toMatchSnapshot();
  });
});
