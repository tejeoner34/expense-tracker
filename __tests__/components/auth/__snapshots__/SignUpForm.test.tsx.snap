// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm should match the snapshot 1`] = `
<DocumentFragment>
  <form
    class="w-full"
  >
    <div
      class="space-y-2"
    >
      <div
        class="grid gap-2"
        data-slot="form-item"
      >
        <label
          class="flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 data-[error=true]:text-destructive-foreground"
          data-error="false"
          data-slot="form-label"
          for=":r6:-form-item"
        >
          Email
        </label>
        <input
          aria-describedby=":r6:-form-item-description"
          aria-invalid="false"
          class="border-input file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive"
          data-slot="form-control"
          id=":r6:-form-item"
          name="email"
          placeholder="mail@example.com"
          value=""
        />
      </div>
      <div
        class="grid gap-2"
        data-slot="form-item"
      >
        <label
          class="flex items-center gap-2 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 data-[error=true]:text-destructive-foreground"
          data-error="false"
          data-slot="form-label"
          for=":r7:-form-item"
        >
          Password
        </label>
        <input
          aria-describedby=":r7:-form-item-description"
          aria-invalid="false"
          class="border-input file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground flex h-9 w-full min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive"
          data-slot="form-control"
          id=":r7:-form-item"
          name="password"
          placeholder="Enter your password"
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-[color,box-shadow] disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-primary text-primary-foreground shadow-xs hover:bg-primary/90 h-9 px-4 py-2 has-[>svg]:px-3 w-full mt-6"
      data-slot="button"
      type="submit"
    >
      Sign up 
    </button>
  </form>
  <div
    class="mx-auto my-4 flex w-full items-center justify-evenly before:mr-4 before:block before:h-px before:flex-grow before:bg-stone-400 after:ml-4 after:block after:h-px after:flex-grow after:bg-stone-400"
  >
    or
  </div>
  <p
    class="text-center text-sm text-gray-600 mt-2"
  >
    If you don't have an account, pleaseÂ 
    <a
      class="text-blue-500 hover:underline"
      href="/sign-in"
    >
      Sign in
    </a>
  </p>
</DocumentFragment>
`;
